riot/venv.py

```diff
@@ -0,0 +1,206 @@
+                try:
+                    run_cmd_venv(
+                        venv_path,
+                        f"pip --disable-pip-version-check install --target {self.target} {self.pkg_str}",
+                        env=env,
+                    )
+                except CmdFailure as e:
+                    raise CmdFailure(
+                        f"Failed to install venv dependencies {self.pkg_str}\n{e.proc.stdout}",
+                        e.proc,
+                    )
+        if self.parent is not None:
+            self.parent.prepare(env, py)
+
+
+@dataclasses.dataclass
+class VenvSpec:
+    """Specifies how to build and run a virtual environment.
+
+    Venvs can be nested to benefit from inheriting from a parent Venv. All
+    attributes are passed down to child Venvs. The child Venvs can override
+    parent attributes with the semantics defined below.
+
+    Example::
+
+        Venv(
+          pys=[3.9],
+          venvs=[
+              Venv(
+                  name="mypy",
+                  command="mypy",
+                  pkgs={
+                      "mypy": "==0.790",
+                  },
+              ),
+              Venv(
+                  name="test",
+                  pys=["3.7", "3.8", "3.9"],
+                  command="pytest",
+                  pkgs={
+                      "pytest": "==6.1.2",
+                  },
+              ),
+          ])
+
+    Args:
+        name (str): Name of the instance. Overrides parent value.
+        command (str): Command to run in the virtual environment. Overrides parent value.
+        pys  (List[float]): Python versions. Overrides parent value.
+        pkgs (Dict[str, Union[str, List[str]]]): Packages and version(s) to install into the virtual env. Merges and overrides parent values.
+        env  (Dict[str, Union[str, List[str]]]): Environment variables to define in the virtual env. Merges and overrides parent values.
+        venvs (List[Venv]): List of Venvs that inherit the properties of this Venv (unless they are overridden).
+    """
+
+    pys: dataclasses.InitVar[t.List[Interpreter._T_hint]] = None
+    pkgs: dataclasses.InitVar[t.Dict[str, t.List[str]]] = None
+    venvs: t.List["VenvSpec"] = dataclasses.field(default_factory=list)
+
+    def __post_init__(self, pys, pkgs):
+        """Normalize the data."""
+        self.pys = [Interpreter(py) for py in to_list(pys)] if pys is not None else []
+        self.pkgs = rm_singletons(pkgs) if pkgs else {}
+
+    def instances(
+        self,
+        pys: t.List[Interpreter] = [],
+        parent_venv: t.Optional[Venv] = None,
+    ) -> t.Generator[Venv, None, None]:
+        pys = pys or self.pys or [None]
+        for py in pys:
+            for ps in expand_specs(self.pkgs):
+                venv = Venv(
+                    py=py,
+                    pkgs=ps,
+                    parent=parent_venv,
+                )
+                if not self.venvs:
```

Comment:
I think the `if` and the `for` loops should be this way (currently it's inside-out). This way one can set the parenting relation between venv instances.